aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -0,0 +1,209 @@
+                    (hist, bin_edges, input, min, max, extract_weight_float);
+            });
+        } else if (weight_dtype == ScalarType::Double) {
+            const double *data_wt = weight.value().data_ptr<double>();
+            auto extract_weight_double = [&data_wt](int64_t i) -> double {
+                return data_wt[i];
+            };
+            AT_DISPATCH_ALL_TYPES(input.scalar_type(), "histogram_out_cpu", [&] {
+                histogram_cpu_contiguous<scalar_t, double, decltype(extract_weight_double), LinearBinEdges>
+                    (hist, bin_edges, input, min, max, extract_weight_double);
+            });
+        } else assert(false);
+
+        return;
+    }
+
+    if (density) {
+        auto extract_weight_double = [](int64_t i) -> double {
+            return 1.0;
+        };
+        AT_DISPATCH_ALL_TYPES(input.scalar_type(), "histogram_out_cpu", [&] {
+            histogram_cpu_contiguous<scalar_t, double, decltype(extract_weight_double), LinearBinEdges>
+                (hist, bin_edges, input, min, max, extract_weight_double);
+        });
+    } else {
+        auto extract_weight_long = [](int64_t i) -> int64_t {
+            return 1L;
+        };
+        AT_DISPATCH_ALL_TYPES(input.scalar_type(), "histogram_out_cpu", [&] {
+            histogram_cpu_contiguous<scalar_t, int64_t, decltype(extract_weight_long), LinearBinEdges>
+                (hist, bin_edges, input, min, max, extract_weight_long);
+        });
+    }
+}
+
+inline void histogram_maybe_trim_input_tensors(
+    Tensor& trimmed_input,
+    Tensor& trimmed_bin_edges,
+    Tensor& trimmed_weight,
+    const Tensor& raw_input,
+    const Tensor& raw_bin_edges,
+    const c10::optional<Tensor>& weight) {
+  // Reuse function from BucketizationUtils handling input and bin_edges
+  searchsorted_maybe_trim_input_tensors(trimmed_input, trimmed_bin_edges, raw_input, raw_bin_edges);
+
+  if (weight.has_value() && !weight.value().is_contiguous()) {
+    TORCH_WARN_ONCE("input weight tensor is non-contiguous, this will lower the performance due to extra data copy "
+        "when converting non-contiguous tensor to contiguous, please use contiguous input value tensor if possible");
+    trimmed_weight = weight.value().contiguous();
+  }
+}
+
+template<bool LinearBinEdges>
+std::tuple<Tensor&, Tensor&>
+histogram_out_cpu_template(const Tensor& self, const Tensor& bins, const Scalar& min, const Scalar& max, const c10::optional<Tensor>& weight, bool density, Tensor& hist, Tensor& bin_edges) {
+    TORCH_CHECK(bins.numel() > 0, "bins tensor should have at least 1 element, but has 0");
+
+    if (hist.numel() == 0) {
+        hist.resize_({bins.numel() - 1});
+    }
+
+    if (bin_edges.numel() == 0) {
+        bin_edges.resize_({bins.numel()});
+    }
+
+    TORCH_CHECK(bins.numel() == hist.numel() + 1, "bins tensor should have 1 more element than hist tensor, but",
```

Comment:
Aren't at least some of these checks redundant with the previous error-checking logic? It seems like there are user-facing error checks in a few places